pkg/ottl/ottlfuncs/func_parse_csv.go

```diff
@@ -0,0 +1,172 @@
+	args, ok := oArgs.(*ParseCSVArguments[K])
+	if !ok {
+		return nil, fmt.Errorf("ParseCSVFactory args must be of type *ParseCSVArguments[K]")
+	}
+
+	if err := args.validate(); err != nil {
+		return nil, fmt.Errorf("invalid arguments: %w", err)
+	}
+
+	delimiter := parseCSVDefaultDelimiter
+	if !args.Delimiter.IsEmpty() {
+		delimiter = []rune(args.Delimiter.Get())[0]
+	}
+
+	// headerDelimiter defaults to the chosen delimter,
+	// since in most cases headerDelimiter == delmiter.
+	headerDelimiter := delimiter
+	if !args.HeaderDelimiter.IsEmpty() {
+		headerDelimiter = []rune(args.HeaderDelimiter.Get())[0]
+	}
+
+	mode := parseCSVDefaultMode
+	if !args.Mode.IsEmpty() {
+		mode = args.Mode.Get()
+	}
+
+	switch mode {
+	case parseCSVModeStrict:
+		return parseCSV(args.Target, args.Header, delimiter, headerDelimiter, false), nil
+	case parseCSVModeLazyQuotes:
+		return parseCSV(args.Target, args.Header, delimiter, headerDelimiter, true), nil
+	case parseCSVModeIgnoreQuotes:
+		return parseCSVIgnoreQuotes(args.Target, args.Header, delimiter, headerDelimiter), nil
+	}
+
+	return nil, fmt.Errorf("unknown mode: %s", mode)
+}
+
+func parseCSV[K any](target, header ottl.StringGetter[K], delimiter, headerDelimiter rune, lazyQuotes bool) ottl.ExprFunc[K] {
+	headerDelimiterString := string([]rune{headerDelimiter})
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		targetStr, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get target: %w", err)
+		}
+
+		headerStr, err := header.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get header: %w", err)
+		}
+
+		if headerStr == "" {
+			return nil, errors.New("headers must not be an empty string")
+		}
+
+		headers := strings.Split(headerStr, headerDelimiterString)
+		fields, err := parseutils.ReadCSVRow(targetStr, delimiter, lazyQuotes)
+		if err != nil {
+			return nil, err
+		}
+
+		headersToFields, err := parseutils.MapCSVHeaders(headers, fields)
+		if err != nil {
+			return nil, fmt.Errorf("map csv headers: %w", err)
+		}
+
+		pMap := pcommon.NewMap()
+		err = pMap.FromRaw(headersToFields)
+		if err != nil {
+			return nil, fmt.Errorf("create pcommon.Map: %w", err)
+		}
+
+		return pMap, nil
+	}
+}
+
+func parseCSVIgnoreQuotes[K any](target, header ottl.StringGetter[K], delimiter, headerDelimiter rune) ottl.ExprFunc[K] {
+	headerDelimiterString := string([]rune{headerDelimiter})
+	delimiterString := string([]rune{delimiter})
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		targetStr, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get target: %w", err)
+		}
+
+		headerStr, err := header.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get header: %w", err)
+		}
+
+		if headerStr == "" {
+			return nil, errors.New("headers must not be an empty string")
+		}
+
+		headers := strings.Split(headerStr, headerDelimiterString)
```

Comment:
I believe a lot of this code is resuable.  I suggest we make a top-level `parseCSV` function that will handle the validation/error checking, then does the `return func(ctx context.Context, tCtx K) (any, error) {` and then inside that function we can get all the data and then make the decision on what type of parsing at the last second.